### DISCUSSION TITLE
how to ajax to '/Login' , I always got error 400?

### DISCUSSION THREAD

**Discussion by scil** [OrchardCMS/OrchardCore#12001] [Q&A] [answered]
update:  ajax has worked. but it's better not to request offical `/Login` by ajax, as offical action returns html response, not well for ajax. I have created custom action to handle user login.

---

Header
```
Request URL: https://localhost:5001/Login
Request Method: POST
Status Code: 400 
```

request palyload
```
Password: "945UIoizh"
UserName: "admin"
rememberMe: true
__RequestVerificationToken:
"x2DJ8C9GTGXpy4RLsPkjnRXSd1OYqyy8F...."
```

I got 400 and `<h1>Your browser sent a request that this server could not understand.</h1>`

it it possible to login by ajax? is it allowd by OrchardCore.User?

**Comment by Skrypt** [comment]
You need to pass the AntiForgeryToken; else it should work.
Are you sure the login URL is /login for your tenant?

**Reply by scil** [reply]
> You need to pass the AntiForgeryToken; else it should work. Are you sure the login URL is /login for your tenant?

Thank you. I did add the anti token.

And I inspect the login form which OC supplied, also the request data which is post to '/Login'

There is no log.

how do I know what's happening deep inside OC?

**Comment by hishamco** [comment]
Please share the AJAX call that you are using

**Reply by scil** [reply]
cshtml
```
<body dir="" class="debug-screens font-hei overflow-x-hidden">
    @Html.AntiForgeryToken()
</body>
```


data
```
data.UserName='admin'
data.Password='xxxxxxxxxxxxxxx78ABC'
data.rememberMe= true;
data.__RequestVerificationToken = document.querySelector('body>input[name="__RequestVerificationToken"]').value
```

ajax
```

              axios.post('/Login', data, {
                  headers:{ 
                      'Accept':'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
                  'Content-type': 'application/x-www-form-urlencoded',
                  }

```

**Comment by ns8482e** [comment, marked as answer]
> I got 400 and `<h1>Your browser sent a request that this server could not understand.</h1>`

@scil you get 400 bad request because the format of data is incorrect. you must send body in format of URL query parameter. i.e. UserName=admin&Password=XYZ 

In ajax call change `data`  using `URLSearchParams` and append each parameter

**Reply by scil** [reply]
safe login need `post`, not `get`.   using url query to transmit passowrd is not good.

**Reply by ns8482e** [reply]
@scil  correct you must send `POST`.  The format of the body needs to be query like `UserName=admin&Password=XYZ` and not JSON as server expects `x-www-form-urlencoded`. 

See https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams -

**Reply by hishamco** [reply]
Agree, the post should passed in the header not URL. Hope to find time to reproduce the issue

**Reply by ns8482e** [reply]
The issue is the format of data sent

**Reply by scil** [reply]
in axios, `x-www-form-urlencoded` is set by this header `    'Content-type': 'application/x-www-form-urlencoded'  `

**Reply by ns8482e** [reply]
Correct ! By setting the header, You are telling server that you are sending  body in urlencoded format, however by  looking at your code it seems that data is still json you need to convert it to url encoded format.  

Using URLSearchParams you can convert json to url encoded query string format

**Reply by scil** [reply]
thank you. your are right.  with axios,  it not work by setting header 'Content-type': 'application/x-www-form-urlencoded' .

now i use `qs`

```
import qs from 'qs'


axios.post('/Login', qs.stringify(data) )
          .then(function () {
            alert(3)
          })
```
